source/computer/macos-setup.rst

```diff
@@ -1,9 +1,160 @@
 macOS 配置指南
 ==============
 
+:本节作者: `王亮 <https://github.com/wangliang1989>`__\ （作者）、
+           `田冬冬 <https://me.seisman.info/>`__\ （作者）、
+           `姚家园 <https://github.com/core-man>`__\ （审稿）
+:最近更新日期: 2021-01-11
+
+----
+
+.. warning::
+
+   本文成稿于 2018 年 1 月，并于 2021 年 1 月做少量改动。
+   但笔者尚未在新电脑上进行测试，无法保证所有命令均正确。欢迎反馈。
+
+第一次启动 Mac 电脑后，经过简单的设置，就得到了一个可供日常使用的 macOS 系统，
+但这样的 macOS 系统尚不能满足日常科研与编程开发的需求，还需要做进一步的配置。
+
+Command Line Tools for Xcode
+----------------------------
+
+Xcode 是 macOS 下的集成开发环境（IDE），类似于 Windows 下的 Microsoft Visual Studio。
+Command Line Tools for Xcode 是 Xcode 的一个子集，其包含了常用的命令行开发工具，
+比如 C 编译器、make、git 等，是安装很多其他软件的基础。
+
+按下 :kbd:`Command+空格`\ ，搜索并打开“Terminal”。在 Terminal 中执行如下命令
+以安装 Command Line Tools for Xcode::
+
+   $ xcode-select --install
+
+在弹出的窗口点击“Install”即可。
+
+此处安装的 Command Line Tools 可能不是最新版，可以到“App Store”的“更新”中或
+“设置”的“软件更新”中查看是否有相关更新。如果有，则升级到最新版。
+
+Homebrew
+--------
+
+`Homebrew <https://brew.sh/index_zh-cn.html>`__ 是 macOS 下的第三方软件包管理器。
+可以用于安装各种常见的软件包、库文件以及字体。日常及科研工作中所需的大多数软件、
+库文件以及字体都可以通过 Homebrew 安装。
+
+要安装 Homebrew，需要将如下命令复制到终端中并执行::
+
+    /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
+
+安装好以后，打开新的 Terminal 便可以使用 ``brew`` 命令。
+Homebrew 的详细用法见官方文档。此处仅列出一些常用的用法::
+
+    # 模糊搜索与 wget 相关的软件
+    $ brew search wget
+
+    # 安装 wget 软件包
+    $ brew install wget
+
+    # 安装 Visual Studio Code
+    # 由于 Visual Studio Code 是带图形界面的软件，因而这里需要使用 --cask 选项
+    $ brew install --cask visual-studio-code
+
+    # 升级某个软件
+    $ brew upgrade xxx
+
+    # 卸载某个软件
+    $ brew uninstall xxx
+
 .. note::
 
-   尚未完成。
+   Homebrew 可能在国内下载速度较慢，可以考虑使用中科大 LUG 提供的 Homebrew 镜像。
+   详细使用方法见：
+
+   - Homebrew 源镜像：http://mirrors.ustc.edu.cn/help/brew.git.html
+   - Homebrew Bottles 源镜像：http://mirrors.ustc.edu.cn/help/homebrew-bottles.html
+   - Homebrew Core 源镜像：http://mirrors.ustc.edu.cn/help/homebrew-core.git.html
+   - Homebrew Cask 源镜像：http://mirrors.ustc.edu.cn/help/homebrew-cask.git.html
+
+编程开发环境
+------------
+
+C
+^
+
+Command Line Tools for Xcode 已经提供了 C 编译器和相关工具，因而无需单独安装
+C编译器。
+
+Java
+^^^^
+
+若要运行 Java 程序，则需要安装 Java 开发环境::
+
+    $ brew install openjdk
```

Comment:
It seems that we need to set openjdk.

```
coreman@Jiayuans-MacBook-Pro research % brew install openjdk
Updating Homebrew...
==> Downloading https://homebrew.bintray.com/bottles/openjdk-15.0.1.big_sur.bottle.tar.gz
==> Downloading from https://d29vzk4ow07wi7.cloudfront.net/6f31366f86a5eacf66673fca9ad647b98b207820f8cfea49a22af596395d3dba?res
######################################################################## 100.0%
==> Pouring openjdk-15.0.1.big_sur.bottle.tar.gz
==> Caveats
For the system Java wrappers to find this JDK, symlink it with
  sudo ln -sfn /usr/local/opt/openjdk/libexec/openjdk.jdk /Library/Java/JavaVirtualMachines/openjdk.jdk

openjdk is keg-only, which means it was not symlinked into /usr/local,
because it shadows the macOS `java` wrapper.

If you need to have openjdk first in your PATH run:
  echo 'export PATH="/usr/local/opt/openjdk/bin:$PATH"' >> ~/.zshrc

For compilers to find openjdk you may need to set:
  export CPPFLAGS="-I/usr/local/opt/openjdk/include"

==> Summary
🍺  /usr/local/Cellar/openjdk/15.0.1: 614 files, 324.9MB
...
```

```
coreman@Jiayuans-MacBook-Pro research % java -version
The operation couldn’t be completed. Unable to locate a Java Runtime.
Please visit http://www.java.com for information on installing Java.
```

```
coreman@Jiayuans-MacBook-Pro research % echo 'export PATH="/usr/local/opt/openjdk/bin:$PATH"' >> ~/.zshrc
coreman@Jiayuans-MacBook-Pro research % source ~/.zshrc 
coreman@Jiayuans-MacBook-Pro research % java -version
openjdk version "15.0.1" 2020-10-20
OpenJDK Runtime Environment (build 15.0.1+9)
OpenJDK 64-Bit Server VM (build 15.0.1+9, mixed mode, sharing)
```

Reply:
It's weird. I have `/usr/bin/java`.

Reply:
I know the reason maybe. When I open a new terminal, it works if I don't add openjdk path  to zshrc.

Reply:
I'm a little confused, do you have `/usr/bin/java`?

Reply:
It's weird. I uninstall openjdk and install it again. Now I can use `java -version` even if the path is added in zshrc.

Reply:
> I'm a little confused, do you have `/usr/bin/java`?

Yes. I have

Reply:
My java is from Sun. How about download java from https://java.com/en/download/apple.jsp

Reply:
I think you should be able to run `java -version` even before you install openjdk.

Reply:
I am a little confused now.
```
coreman@Jiayuans-MacBook-Pro research % which java
/usr/local/opt/openjdk/bin/java
```

Reply:
> I am a little confused now.
> 
> ```
> coreman@Jiayuans-MacBook-Pro research % which java
> /usr/local/opt/openjdk/bin/java
> ```

maybe you install two java

Reply:
```
coreman@Jiayuans-MacBook-Pro research % brew uninstall openjdk       
Uninstalling /usr/local/Cellar/openjdk/15.0.1... (614 files, 324.9MB)
coreman@Jiayuans-MacBook-Pro research % java -version
The operation couldn’t be completed. Unable to locate a Java Runtime.
Please visit http://www.java.com for information on installing Java.
```

Reply:
> My java is from Sun. How about download java from https://java.com/en/download/apple.jsp
> ![image](https://user-images.githubusercontent.com/12059719/104560041-43e7cf00-5680-11eb-8895-97999f90f560.png)

Your java is very old, perhaps you don't upgrade your command line tools for code?

> I am a little confused now.
> 
> ```
> coreman@Jiayuans-MacBook-Pro research % which java
> /usr/local/opt/openjdk/bin/java
> ```

Run `brew list` to check if openjdk is already uninstalled.

Reply:
> > coreman@Jiayuans-MacBook-Pro research % which java
> > /usr/local/opt/openjdk/bin/java
> > ```
> 
> Run `brew list` to check if openjdk is already uninstalled.

```
coreman@Jiayuans-MacBook-Pro research % brew uninstall openjdk
Uninstalling /usr/local/Cellar/openjdk/15.0.1... (614 files, 324.9MB)
coreman@Jiayuans-MacBook-Pro research % brew list
coreman@Jiayuans-MacBook-Pro research % java -version
The operation couldn’t be completed. Unable to locate a Java Runtime.
Please visit http://www.java.com for information on installing Java.
```

Reply:
After uninstalling openjdk, could you please run `which java`, and also see if `/usr/bin/java` is a plaintext file.

Reply:
> After uninstalling openjdk, could you please run `which java`

```
coreman@Jiayuans-MacBook-Pro research % brew uninstall openjdk       
Uninstalling /usr/local/Cellar/openjdk/15.0.1... (614 files, 324.9MB)
coreman@Jiayuans-MacBook-Pro research % which java
/usr/bin/java
```

> and also see if `/usr/bin/java` is a plaintext file.

No, it is binary.

Reply:
I have the feeling that, my `/usr/bin/java` works because I installed the full Xcode, not just the command line tools for code.

We should have checked `gcc`, `gfortran`, `java` et al. before install CLT and homebrew.

Reply:
Okay, I think I made a mistake above. When I find I cannot run java, I add the openjkd path to zshrc:
```
echo 'export PATH="/usr/local/opt/openjdk/bin:$PATH"' >> ~/.zshrc
```

Then I did some commands in the same terminal and some commands in others.

Now I already deleted openjdk path in zshrc and open a new terminal to test.


```
coreman@Jiayuans-MacBook-Pro research % brew uninstall openjdk
Uninstalling /usr/local/Cellar/openjdk/15.0.1... (614 files, 324.9MB)
coreman@Jiayuans-MacBook-Pro research % ls /usr/local/Cellar 
coreman@Jiayuans-MacBook-Pro research % which java
/usr/bin/java
coreman@Jiayuans-MacBook-Pro research % java -version
The operation couldn’t be completed. Unable to locate a Java Runtime.
Please visit http://www.java.com for information on installing Java.

coreman@Jiayuans-MacBook-Pro research % echo $PATH
/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
```

```
coreman@Jiayuans-MacBook-Pro research % brew install openjdk
==> Downloading https://homebrew.bintray.com/bottles/openjdk-15.0.1.big_sur.bottle.tar.gz
Already downloaded: /Users/coreman/Library/Caches/Homebrew/downloads/d1e3ece4af1d225bc2607eaa4ce85a873d2c6d43757ae4415d195751bc431962--openjdk-15.0.1.big_sur.bottle.tar.gz
==> Pouring openjdk-15.0.1.big_sur.bottle.tar.gz
==> Caveats
For the system Java wrappers to find this JDK, symlink it with
  sudo ln -sfn /usr/local/opt/openjdk/libexec/openjdk.jdk /Library/Java/JavaVirtualMachines/openjdk.jdk

openjdk is keg-only, which means it was not symlinked into /usr/local,
because it shadows the macOS `java` wrapper.

If you need to have openjdk first in your PATH run:
  echo 'export PATH="/usr/local/opt/openjdk/bin:$PATH"' >> ~/.zshrc

For compilers to find openjdk you may need to set:
  export CPPFLAGS="-I/usr/local/opt/openjdk/include"

==> Summary
🍺  /usr/local/Cellar/openjdk/15.0.1: 614 files, 324.9MB
coreman@Jiayuans-MacBook-Pro research % echo $PATH
/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
coreman@Jiayuans-MacBook-Pro research % java -version
The operation couldn’t be completed. Unable to locate a Java Runtime.
Please visit http://www.java.com for information on installing Java.

coreman@Jiayuans-MacBook-Pro research % which java
/usr/bin/java
```

Reply:
I probably downloaded the jdk from this site https://www.oracle.com/java/technologies/javase-downloads.html long time ago. That's why my `/usr/bin/java` works.

Reply:
Here are my understanding about the jdk.

There are two ways to make java work on macOS:

1. Download the java pacakge from https://www.oracle.com/java/technologies/javase-downloads.html or https://www.java.com/en/download/ and install it (I actually don't know the differences among so many java versions).
2. Install openjdk via homebrew, and run this command (no need to change PATH):
```
sudo ln -sfn /usr/local/opt/openjdk/libexec/openjdk.jdk /Library/Java/JavaVirtualMachines/openjdk.jdk
```

Reply:
Yes, now we can be sure that java cannot run if OpenJDK is not installed for a new macOS.

If we install OpenJDK but do not add its path to $PATH, the java is still the one in `/usr/local/bin`.

If we add OpenJDK path to `.zshrc`, java can run but it is not the one in `/usr/local/bin`, instead
it is in `usr/local/opt/openjdk/bin`.

Reply:
> ```
> sudo ln -sfn /usr/local/opt/openjdk/libexec/openjdk.jdk /Library/Java/JavaVirtualMachines/openjdk.jdk
> ```

Yes, in this way, the java we used is still from macOS, not openJDK, while $PATH is not changed.

Reply:
> We should have checked `gcc`, `gfortran`, `java` et al. before install CLT and homebrew.

I think we somehow kown how `java` works now.

After I install CLT and homebrew, `gfortran` still does not exist.

I don't know if `gcc` exists before CLT and homebrew is installed.

Reply:
As I understand it, `/usr/bin/java` provided by macOS is just a wrapper. The java wrapper tries to find openjdk in the system path `/Library/Java/JavaVirtualMachines/openjdk.jdk`. If found, then you can run `java`. 

Homebrew installs openjdk to `/usr/local`, so that the java wrapper cannot find it, and we have to make the symblink.

Reply:
> As I understand it, `/usr/bin/java` provided by macOS is just a wrapper. The java wrapper tries to find openjdk in the system path `/Library/Java/JavaVirtualMachines/openjdk.jdk`. If found, then you can run `java`.
> 
> Homebrew installs openjdk to `/usr/local`, so that the java wrapper cannot find it, and we have to make the symblink.

I think so.